cmip6_downscaling/methods/gard/tasks.py

```diff
@@ -0,0 +1,240 @@
+        + str(ytrain_path)
+        + str(xpred_path)
+        + run_parameters.run_id_hash
+        + str(dim)
+    )
+    target = intermediate_dir / 'gard_fit_and_predict' / ds_hash
+
+    if use_cache and zmetadata_exists(target):
+        print(f'found existing target: {target}')
+        return target
+
+    kws = default_none_kwargs(run_parameters.bias_correction_kwargs, copy=True)
+    print(xtrain_path)
+    print(ytrain_path)
+    print(xpred_path)
+    # load in datasets
+    xtrain = xr.open_zarr(xtrain_path)
+    ytrain = xr.open_zarr(ytrain_path)
+    xpred = xr.open_zarr(xpred_path)
+    print(xtrain.chunks)
+    print(ytrain.chunks)
+    print(xpred.chunks)
+
+    # make sure you have the variables you need in obs
+    for v in xpred.data_vars:
+        assert v in ytrain.data_vars
+
+    # data transformation (this wants full-time chunking)
+    # transformed_obs is for the training period
+    transformed_obs = bias_correct_obs_by_method(
+        da_obs=ytrain, method=run_parameters.bias_correction_method, bc_kwargs=kws
+    ).to_dataset(dim="variable")
+
+    # transformed_gcm is for the prediction period
+    transformed_gcm = bias_correct_gcm_by_method(
+        gcm_train=xtrain,
+        obs_train=ytrain,
+        gcm_predict=xpred,
+        method=run_parameters.bias_correction_method,
+        bc_kwargs=kws,
+    ).to_dataset(dim="variable")
+
+    # model definition
+    model = PointWiseDownscaler(
+        model=get_gard_model(run_parameters.model_type, run_parameters.model_params), dim=dim
+    )
+
+    print(transformed_gcm.sel(time=run_parameters.train_period.time_slice))
+    print(transformed_obs[run_parameters.variable])
+    # model fitting
+    model.fit(
+        transformed_gcm.sel(time=run_parameters.train_period.time_slice).assign_coords(
+            {"time": transformed_obs.time.values}
+        ),
+        transformed_obs[run_parameters.variable],
+    )
+
+    # model prediction
+    out = model.predict(transformed_gcm).to_dataset(dim='variable')
+    out.to_zarr(target)
+    return out
+
+
+@task(log_stdout=True)
+def postprocess(
+    model_output_path: UPath,
+    run_parameters: RunParameters,
+) -> UPath:
+    """
+    Add perturbation to the mean prediction of GARD to more accurately represent extreme events. The perturbation is
+    generated with the prediction error during model fit scaled with a spatio-temporally correlated random field.
+
+    Parameters
+    ----------
+    model_output_path : UPath
+        GARD model prediction output. Should contain three variables: pred (predicted mean), prediction_error
+        (prediction error in fit), and exceedance_prob (probability of exceedance for threshold)
+    run_parameters : RunParameters
+        Model parameter dictionary
+
+    Returns
+    -------
+    downscaled : UPath
+        Final downscaled output
+    """
+    ds_hash = str_to_hash(str(model_output_path) + run_parameters.run_id_hash)
+    target = results_dir / 'daily' / ds_hash
```

Comment:
We should think about how we want the core science results to be organized. I think this is a place where the `run_id` is going to be important.